### PR TITLE
Define feature types and access/iterators for file

### PR DESCRIPTION
**Pull request by Ipiano**

### CODE REVIEW

File: include/shpxx/shpfile_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    return FeatureT(*object);
+}
+
+template<shpxx::concepts::HasCompatibilityCheck CompatibilityCheckT>
+void shpfile_t::throw_if_incompatible() const
+{
+    if (!CompatibilityCheckT::is_compatible(type()))
+    {
+        throw incompatible_feature_type_error_t(
+            "specified feature type is not compatible with file");
```
Comment (Notgnoshi) [inline include/shpxx/shpfile_impl.hpp]: Error messages are best if they contain more specific information. E.g.,
```c++
"specified feature type {feature_type} is not compatible with file's feature type {file_type}"
```

But maybe this is gross to do in this case?

Comment (Ipiano) [reply include/shpxx/shpfile_impl.hpp]: I'd have to go through and define a bunch of string conversions, but yes, that is possible

File: include/shpxx/shpfile.hpp [hunk 2]
```diff
@@ -68,6 +105,35 @@ class shpfile_t
...
+    //! converted to. This must be compatible with the shape_type returned by
+    //! type()
+    //!
+    //! @throw incompatible_feature_type_error_t if FeatureT is not compatible
+    //! with the type of features indicated by the file header
+    //!
+    //! @return A new feature iterator which can be used to access the features
+    //! of the file as a specific feature type
+    template<shpxx::concepts::IsFeature FeatureT>
+    [[nodiscard]] const_feature_iterator_t<FeatureT> begin() const;
```
Comment (Notgnoshi) [inline include/shpxx/shpfile.hpp:122]: Should this be `cbegin()` and `cend()`?

Comment (Ipiano) [reply include/shpxx/shpfile.hpp:122]: Should probably have aliases for those. afaik, convention is for `cbegin() const` and `begin() const` to be identical and/or defer one to the other

Comment (Ipiano) [reply include/shpxx/shpfile.hpp:122]: Added

File: src/shape_type.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+#include <string>
+
+namespace shpxx {
+
+std::string to_string(shape_type type) noexcept
+{
+    switch (type)
+    {
+    case shape_type::null:
+        return "null";
```
Comment (Ipiano) [inline src/shape_type.cpp:12]: Should I make all these match the enum exactly? e.g. `shpxx::shape_type::null`, `shpxx::shape_type::point`

**Review by Notgnoshi — APPROVED** [review]

**Pull request merged**